stratum/hal/lib/common/phaldb_service.cc

```diff
@@ -0,0 +1,454 @@
+    if (req_path.entries_size() == 0) {
+        RETURN_ERROR(ERR_INVALID_PARAM) << "No Path";
+    }
+
+    // Create Attribute DB Path
+    ::stratum::hal::phal::PathEntry entry;
+    for (const auto& ent : req_path.entries()) {
+        entry.name = ent.name();
+        entry.index = ent.index();
+        entry.indexed = ent.indexed();
+        entry.all = ent.all();
+        entry.terminal_group = ent.terminal_group();
+        path.push_back(entry);
+    }
+
+    return path;
+}
+
+}  // namespace
+
+::grpc::Status PhalDBService::Get(::grpc::ServerContext* context,
+    const ::stratum::hal::phal::GetRequest* req,
+    ::stratum::hal::phal::GetResponse* resp) {
+
+  RETURN_IF_NOT_AUTHORIZED(auth_policy_checker_, PhalDBService, Get, context);
+
+  ::util::Status status;
+  ::util::StatusOr<::stratum::hal::phal::Path> result;
+
+  // Convert to PhalDB Path
+  switch (req->query_case()) {
+    case ::stratum::hal::phal::GetRequest::kStr: {
+        result = ParseQuery(req->str());
+        break;
+    }
+    case ::stratum::hal::phal::GetRequest::kPath: {
+        result = ToPhalDBPath(req->path());
+        break;
+    }
+    default:
+        return ::grpc::Status(
+            ::grpc::StatusCode::INVALID_ARGUMENT,
+            "Invalid query in Get request.");
+  }
+
+  if (!result.ok()) return ToGrpcStatus(result.status(), {});
+
+  auto path = result.ConsumeValueOrDie();
+
+  // Issue the get
+  auto adapter =
+    absl::make_unique<::stratum::hal::phal::Adapter>(phal_interface_);
+  auto phaldb_res = adapter->Get({path});
+  if (phaldb_res.ok()) {
+      auto phaldb_resp = std::move(phaldb_res.ConsumeValueOrDie());
+      *resp->mutable_phal_db() = *phaldb_resp;
+  }
+
+  // Convert to grpc status and return
+  return ToGrpcStatus(phaldb_res.status(), {});
+}
+
+::grpc::Status PhalDBService::Set(::grpc::ServerContext* context,
+    const ::stratum::hal::phal::SetRequest* req,
+    ::stratum::hal::phal::SetResponse* resp) {
+
+  RETURN_IF_NOT_AUTHORIZED(auth_policy_checker_, PhalDBService, Set, context);
+
+  if (!req->updates_size()) return ::grpc::Status::OK;  // Nothing to do.
+
+  ::util::Status status = ::util::OkStatus();
+  std::vector<::util::Status> results = {};
+  stratum::hal::phal::AttributeValueMap attrs;
+
+  // Spin thru each update
+  for (int i=0; i < req->updates_size(); i++) {
+    ::util::StatusOr<::stratum::hal::phal::Path> attr_res;
+
+    // Get the update
+    const auto& update = req->updates(i);
+
+    // Convert to PhalDB Path
+    switch (update.query_case()) {
+      case ::stratum::hal::phal::Update::kStr: {
+        attr_res = ParseQuery(update.str());
+        break;
+      }
+      case ::stratum::hal::phal::Update::kPath: {
+        attr_res = ToPhalDBPath(update.path());
+        break;
+      }
+      default:
+        attr_res = MAKE_ERROR(ERR_INVALID_PARAM) << "Invalid update query";
+    }
+
+    if (!attr_res.ok()) {
+        LOG(ERROR) << "Set update " << std::to_string(i) << " failed: "
```

Comment:
1. `int` converts automatically when stream-redirected

2. Instead of logging the position of the update, log the update itself: `update.ShortDebugString`

Reply:
good suggestion, done